crates/wasi/src/lib.rs

```diff
@@ -28,3 +30,38 @@ fn wasi_proc_exit(status: i32) -> Result<(), Trap> {
         ))
     }
 }
+
+/// Creates a new [`Linker`], similar to `Linker::new`, and initializes it
+/// with WASI exports.
+pub fn wasi_linker(
+    store: &Store,
+    preopen_dirs: &[(String, File)],
+    argv: &[String],
+    vars: &[(String, String)],
+) -> Result<Linker> {
+    let mut linker = Linker::new(store);
+
+    let mut cx = WasiCtxBuilder::new();
+    cx.inherit_stdio().args(argv).envs(vars);
+
+    for (name, file) in preopen_dirs {
+        cx.preopened_dir(file.try_clone()?, name);
+    }
+
+    let cx = cx.build()?;
+    let wasi = Wasi::new(linker.store(), cx);
+    wasi.add_to_linker(&mut linker)?;
+
+    let mut cx = old::snapshot_0::WasiCtxBuilder::new();
+    cx.inherit_stdio().args(argv).envs(vars);
+
+    for (name, file) in preopen_dirs {
+        cx.preopened_dir(file.try_clone()?, name);
+    }
```

Comment:
Why do we preopen dirs twice?

Reply:
Once for snapshot1 and once for snapshot0. I've now added a comment about this.